src/Common/tests/System/Net/Http/LoopbackServer.cs

```diff
@@ -454,32 +463,80 @@ public override async Task<HttpRequestData> HandleRequestAsync(HttpStatusCode st
 
-            // Convert header lines to key/value pairs
-            // Skip first line since it's the status line
-            foreach (var line in headerLines.Skip(1))
-            {
-                int offset = line.IndexOf(':');
-                string name = line.Substring(0, offset);
-                string value = line.Substring(offset + 1).TrimStart();
-                requestData.Headers.Add(new HttpHeaderData(name, value));
-            }
+                // Parse method and path
+                string[] splits = headerLines[0].Split(' ');
+                requestData.Method = splits[0];
+                requestData.Path = splits[1];
+
+                // Convert header lines to key/value pairs
+                // Skip first line since it's the status line
+                foreach (var line in headerLines.Skip(1))
+                {
+                    int offset = line.IndexOf(':');
+                    string name = line.Substring(0, offset);
+                    string value = line.Substring(offset + 1).TrimStart();
+                    requestData.Headers.Add(new HttpHeaderData(name, value));
+                }
+
+                if (requestData.Method != "GET")
+                {
+                    if (requestData.GetHeaderValueCount("Content-Length") != 0)
+                    {
+                        int contentLength = Int32.Parse(requestData.GetSingleHeaderValue("Content-Length"));
+                        byte[] buffer = new byte[contentLength];
+                        int offset = 0;
+
+                        while (contentLength != 0)
+                        {
+                            int length = await connection.Reader.BaseStream.ReadAsync(buffer, offset, contentLength);
+                            contentLength -= length;
+                            offset += length;
+                        }
+
+                        requestData.AddBodyData(buffer);
+                    }
+                    else if (requestData.GetHeaderValueCount("Transfer-Encoding") != 0 && requestData.GetSingleHeaderValue("Transfer-Encoding") == "chunked")
+                    {
+                        while (true)
+                        {
+                            string chunkHeader = await connection.Reader.ReadLineAsync();
+                            if (chunkHeader == null)
+                            {
+                                // Some tests do not send complete body.
+                                break;
+                            }
+                            int contentLength = int.Parse(chunkHeader, System.Globalization.NumberStyles.HexNumber);
+                            if (contentLength == 0)
+                            {
+                                // Last chunk
+                                break;
+                            }
+                            char[] buffer = new char[contentLength];
```

Comment:
Why is this char instead of byte?

Reply:
I did not find good method in StreamReader. When I used StreamReader.BaseStream I was getting incomplete data. It seems like the StreamReader can consume some and store them in internal buffer. 
If you or @stephentoub  have any specific suggestion, please let me know.

Reply:
It turned out is was bad idea to mic StreamReader.ReadAsync/ReadLine with StreamReader.BaseStream.ReadAsync. It would be nice to get byte[] back using IO on underlying stream but so far I did not find reliable way how to make it happen.
So for now I always use  ReadAsync returning char[] and converting to byte[].

Reply:
We really shouldn't be using StreamReader to implement the loopback server. I forgot we were doing this. We should change this so we don't use StreamReader, but that's a fair amount of work to do, unfortunately.

What's driving you to add support for reading the request body here? It would be nice to have this, but I'm not sure why it's needed as part of this PR specifically. And if we are going to add this support, it seems like we should do it properly, rather than hack it.

Reply:
The test I was trying to fix and use was using POST. 
It does not have to but I thought it would be nice to get it in so we can keep progressing on fixing tests. 
In this particular case I can change it to GET or disable this for HTTP/1.x 

Let me know what direction do you want to take @geoffkizer